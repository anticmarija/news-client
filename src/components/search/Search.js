import React, { useState, useEffect } from "react";
import { useSelector } from "react-redux";
import { findActiveCountryName } from "../../utils/helpers";
import config from "../../config";
import useDebounce from "../../hooks/use-debounce";
import Thumbnail from "../thumbnail/Thumbnail";
import { NewsWrapperStyle, MainHeaderStyle } from "../../style/Shared.style";
import {
  SearchWrapperStyle,
  SearchInputStyle,
  SearchingStatusStyle,
} from "./Search.style";

const apiHandler = async (activeCountry, searchingTerm) => {
  const response = await fetch(
    `${config.getApiBaseUrl()}/top-headlines?country=${activeCountry}&q=${searchingTerm}&apiKey=${config.getApiKey()}`
  );
  const { articles } = await response.json();
  return articles;
};

const Search = () => {
  let { activeCountry, supportedCountries } = useSelector(
    (state) => state.countries
  );
  const [searchingStatus, setSearchingStatus] = useState(false);
  const [searchingTerm, setSearchingTerm] = useState("");
  const [resultArticles, setResultArticles] = useState([]);

  const [isNthMatched, setIsNthMatched] = useState(false);

  const debouncedSearchTerm = useDebounce(searchingTerm, 500);

  useEffect(() => {
    if (debouncedSearchTerm) {
      setSearchingStatus(true);
      setIsNthMatched(false);
      const asyncApiCall = async () => {
        const articles = await apiHandler(activeCountry, debouncedSearchTerm);

        setSearchingStatus(false);
        setResultArticles(articles);

        setIsNthMatched(articles.length === 0);
      };

      asyncApiCall();
    } else {
      setResultArticles([]);
    }
  }, [debouncedSearchTerm, activeCountry]);

  const countryName = findActiveCountryName(supportedCountries, activeCountry);

  return (
    <SearchWrapperStyle>
      <MainHeaderStyle>
        Search top news from {countryName} by term:
      </MainHeaderStyle>
      <SearchInputStyle
        placeholder="Search term..."
        onChange={(event) => setSearchingTerm(event.target.value)}
      ></SearchInputStyle>
      <SearchingStatusStyle show={searchingStatus || isNthMatched}>
        {searchingStatus ? "Searching..." : isNthMatched && "Nothing matches!"}
      </SearchingStatusStyle>
      <NewsWrapperStyle>
        {resultArticles.map((article) => (
          <Thumbnail
            key={`${article.title}_${article.publishedAt}`}
            article={article}
          />
        ))}
      </NewsWrapperStyle>
    </SearchWrapperStyle>
  );
};

export default Search;
