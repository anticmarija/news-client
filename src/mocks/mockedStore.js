import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

const middlewares = [thunk];
export const initialState = {
  topNews: {},
  countries: {
    supportedCountries: [{ countryId: "gb", countryName: "Gb" }],
  },
  categories: {},
  article: {
    activeArticle: {
      title: "active",
      urlToImage:
        "https://i.dailymail.co.uk/1s/2020/06/17/16/29719992-0-image-a-23_1592408042536.jpg",
      description: "desc",
    },
  },
};

export const mockedStore = (state = initialState) => {
  const mockStore = configureStore(middlewares);
  return mockStore(state);
};
