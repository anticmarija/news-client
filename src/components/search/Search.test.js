import React from "react";
import Search from "./Search";
import { mockedStore } from "../../mocks/mockedStore";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

describe("Test Search component", () => {
  it("should render correctly", () => {
    const store = mockedStore();

    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Search />
        </MemoryRouter>
      </Provider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
