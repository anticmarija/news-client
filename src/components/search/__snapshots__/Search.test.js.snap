// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Search component should render correctly 1`] = `
<div
  class="Searchstyle__SearchWrapperStyle-sc-15swxjh-0 eEbpPg"
>
  <h1
    class="Sharedstyle__MainHeaderStyle-sc-1a3bnp9-0 chhIjm"
  >
    Search top news from 
    Gb
     by term:
  </h1>
  <input
    class="Searchstyle__SearchInputStyle-sc-15swxjh-1 eqKNsS"
    data-testid="search-input"
    placeholder="Search term..."
  />
  <div
    class="Searchstyle__SearchingStatusStyle-sc-15swxjh-2 dzlCMZ"
  />
  <div
    class="Sharedstyle__NewsWrapperStyle-sc-1a3bnp9-1 jTferJ"
  />
</div>
`;
